src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java

```diff
@@ -1987,6 +1987,223 @@ public StarlarkExecutionResult execute(
+          (in bytes, default 64 MiB).
+
+          <p>This method returns a <code>wasm_exec_result</code> structure containing
+          the function's return code (in field <code>return_code</code>) and output
+          buffer (in field <code>output</code>). The return code will be negative
+          if execution timed out.
+          """,
+      useStarlarkThread = true,
+      enableOnlyWithFlag = BuildLanguageOptions.EXPERIMENTAL_REPOSITORY_CTX_EXECUTE_WASM,
+      parameters = {
+        @Param(
+            name = "module",
+            positional = true,
+            named = true,
+            allowedTypes = {
+              @ParamType(type = String.class),
+              @ParamType(type = Label.class),
+              @ParamType(type = StarlarkPath.class),
+              @ParamType(type = StarlarkWasmModule.class)
+            },
+            doc =
+                """
+                Path of the WebAssembly module to execute, or a <code>wasm_module</code>
+                loaded by <a href="#load_wasm"><code>load_wasm</code></a>.
+                """),
+        @Param(
+            name = "function",
+            positional = true,
+            named = true,
+            doc = "The name of the function to execute"),
+        @Param(
+            name = "input",
+            positional = false,
+            named = true,
+            doc = "The content of the input buffer."),
+        @Param(
+            name = "timeout",
+            defaultValue = "600",
+            positional = false,
+            named = true,
+            doc = "Execution timeout in seconds (default is 600 seconds)."),
+        @Param(
+            name = "memory_limit",
+            defaultValue = "67108864", // 64 MiB
+            positional = false,
+            named = true,
+            doc = "Memory limit in bytes (default is 64 MiB"),
+        @Param(
+            name = "watch",
+            defaultValue = "'auto'",
+            positional = false,
+            named = true,
+            doc =
+                """
+                Whether to <a href="#watch">watch</a> the file. Can be the string 'yes', 'no', \
+                or 'auto'. Passing 'yes' is equivalent to immediately invoking the \
+                <a href="#watch"><code>watch()</code></a> method; passing 'no' does not \
+                attempt to watch the file; passing 'auto' will only attempt to watch the \
+                file when it is legal to do so (see <code>watch()</code> docs for more \
+                information.
+                """)
+      })
+    public StarlarkWasmExecutionResult executeWasm(
+      Object pathOrModule,
+      String function,
+      String input,
+      StarlarkInt timeoutI,
+      StarlarkInt memLimitI,
+      String watch,
+      StarlarkThread thread)
+      throws EvalException, RepositoryFunctionException, InterruptedException {
+    StarlarkPath path = null;
+    StarlarkWasmModule wasmModule = null;
```

Comment:
If you have many instances of the same repo rule, it seems like they would all need to load (and potentially compile) the same shared WASM helper. We don't necessarily have to optimize this now, but make sure that we don't have ideas for a design that avoids this or, at least, that the current approach doesn't preclude optimizations (say cache modules by hash of the WASM module).

Reply:
The `com.dylibso.chicory.wasm.WasmModule` type is immutable, so it should be safe to cache the `com.dylibso.chicory.wasm.Parser.parse(moduleContent)` call based on the content (or content checksum) of the loaded module.

A `StarlarkWasmModule` is a `WasmModule` plus a few config parameters (name of the allocation function, maybe more in future?) so it should be cheap to create as many copies as needed if the `WasmModule` parsing is cached.

Reply:
Thanks for confirming that `WasmModule` is safe to share. Generic caching of `WasmModule` doesn't appear to be absolutely straightforward since the modules can be large and the lifetimes are not obvious. But I also don't see an API that would solve this. Let's just wait until it becomes a problem in practice. :-)